docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,137 @@
+# MeshService migration
+
+* Status: accepted
+
+## Context and Problem Statement
+
+This MADR addresses the switch from not having MeshServices at all to MeshServices
+being automatically created, either from `Services` in Kubernetes or from
+`Dataplanes` in Universal.
+
+When upgrading from 2.8 to 2.9, nothing at the data plane level should change.
+But MeshService behaves differently in several different ways.
+
+### Policy
+
+If for example we have a policy targeted at:
+
+```
+spec:
+  targetRef:
+    kind: MeshService
+    name: demo-app_kuma-demo_svc_5000
+```
+
+when switching over to MeshService, this policy will no longer apply.
+
+### Local traffic
+
+Because the `MeshServices` we generate have selectors equivalent to the `Service`,
+they will select the same `Dataplanes` that previously had the `kuma.io/service`
+corresponding to the `Service` and thus, absent any policies, traffic flows to
+the same endpoints and Envoy has the same cluster configuration.
+
+### Cross-zone traffic
+
+One of the fundamental changes with `MeshService` is that traffic to a service
+is no longer load balanced across all zones.
+
+We need to take measures to prevent this change from happening on upgrade.
+
+In particular with Kubernetes, it's important to keep in mind that after
+MeshService is enabled, _any_ requests to the Kubernetes IP/via Kube DNS
+go to the same outbound as to MeshService hostnames.
+
+#### Status quo
+
+- On Kubernetes there are two outbound listeners:
+  - Traffic to the Kube DNS name/Cluster IP is load-balanced cross-zone
+  - Traffic to the `<kuma.io/service>.mesh`/VIP is load-balanced cross-zone
+
+#### `MeshService`
+
+- We now have one outbound listener:
+  - Traffic to the generated hostname of the `MeshService` goes through the Cluster IP
+  - This cluster IP is no longer cross-zone load balanced
+
+## Considered Options
+
+* Policies applied to `demo-app_kuma-demo_svc_5000` should also apply to
+  the `MeshService {name: demo-app, namespace: kuma-demo, spec.ports[port == 5000]}`.
+* For cross-zone:
+  * don't generate MeshService by default
+  * have a setting to disable using MeshServices
+  * change the default behavior of MeshService from single-zone load balancing to cross-zone.
+    Users would need to manually switch over to local zone load-balancing after having
+    potentially migrated cross-zone services to `MeshMultiZoneService`.
+  * instead of changing the behavior of MeshService, instead also generate
+    MeshMultiZoneService by default, which would reintroduce the cross-zone load balancing behavior.
+
+## Decision Outcome
+
+* Policies applied to `demo-app_kuma-demo_svc_5000` should also apply to
+  the `MeshService `{name: demo-app, namespace: kuma-demo, spec.ports[port == 5000]}`.
+
```

Comment:
Do we have the decision for the `cross-zone`?

Reply:
not yet, i'll add a tentative decision